File: scripts/sample02_04.m
%[text] # Sample 2-4
%[text] ## 画像データの入出力
%[text] ヒストグラム均等化 
%[text] 画像処理特論
%[text] 村松 正吾 
%[text] 動作確認: MATLAB R2023a
%[text] ## Input and output of images
%[text] Histogram equalization
%[text] Advanced Topics in Image Processing
%[text] Shogo MURAMATSU
%[text] Verified: MATLAB R2023a
%%
%[text] ### サンプル画像の準備
%[text] (Preparation of sample image)
%[text] 本サンプルで利用する画像データを収めたdata フォルダにパスをとおす。
%[text] Create a path to the data folder that contains images used in this sample.
addpath('./data')
close
% Reading original image
I = im2double(rgb2gray(imread('firenzeRgb.jpg')));
figure(1)
imshow(I)
title('Original')
figure(2)
imhist(I)
%%
%[text] ### ヒストグラム均等化前の加工
%[text] (Process before histogram equalization)
%[text]{"align":"center"} $v=T\_1(x)= x^\\gamma $
%[text]{"align":"center"} $y=T\_2(v)= \\frac{1}{2}(\\mathrm{sign}(2v-1)|2v-1|^{10^{-\\alpha}}+1)$
% Definition of process
gamma = 1 %[control:slider:52e5]{"position":[9,10]}
alpha = 0 %[control:slider:74f2]{"position":[9,10]}
T1 = @(x) x.^gamma;
T2 = @(v) 0.5*(sign(2.0*v-1.0).*abs(2*v-1.0).^(10^(-alpha))+1.0);
Tp = @(x) T2(T1(x));
figure(3)
fplot(Tp,[0,1])
axis square
grid on
% Preprocessing for histogram equalization
J = Tp(I);
figure(4)
imshow(J)
title('Preprocessed image for histogram equalization')
figure(5)
imhist(J)
%%
%[text] ### ヒストグラム均等化
%[text] (Histgram equalization)
% Power law conversion with IMADUST function
K = histeq(J);
figure(6)
imshow(K)
title('Result of histogram equalization')
figure(7)
imhist(K)
%%
%[text] © Copyright, Shogo MURAMATSU, All rights reserved.

%[appendix]{"version":"1.0"}
%---
%[metadata:view]
%   data: {"layout":"inline","rightPanelPercent":40}
%---
%[control:slider:52e5]
%   data: {"defaultValue":1,"label":"gamma","max":4,"min":0,"run":"SectionToEnd","runOn":"ValueChanging","step":0.1}
%---
%[control:slider:74f2]
%   data: {"defaultValue":0,"label":"alpha","max":4,"min":-4,"run":"SectionToEnd","runOn":"ValueChanging","step":0.1}
%---

File: python/sample02_04.py
# -*- coding: utf-8 -*-
#%% Sample 2-4
#%% 画像データの入出力
# ヒストグラム均等化 
# 
# 画像処理特論
# 
# 村松 正吾 
# 
# 動作確認: Python 3.7, PyTorch 1.8
#%% Input and output of images
# Histogram equalization
# 
# Advanced Topics in Image Processing
# 
# Shogo MURAMATSU
# 
# Verified: Python 3.7, PyTorch 1.8
from PIL import Image, ImageOps
import requests
import torch
import torchvision
from matplotlib import pyplot as plt
im2double = torchvision.transforms.ConvertImageDtype(torch.double)
rgb2gray = torchvision.transforms.Grayscale()
totensor = torchvision.transforms.ToTensor()
topilimg = torchvision.transforms.ToPILImage()
locs = torch.linspace(0,1,256).numpy()
imhist = lambda x: plt.stem(locs,torch.histc(x,bins=256,min=0.,max=1.),markerfmt='None',use_line_collection=True)

#%% サンプル画像の準備
# (Preparation of sample image)
# 
# 本サンプルで利用する画像データを収めたdata フォルダにパスをとおす。
# 
# Create a path to the data folder that contains images used in this sample.

# Reading original image
I = im2double(rgb2gray(totensor(Image.open('./data/firenzeRgb.jpg'))))
#url = 'https://github.com/msiplab/AtipWork/raw/master/data/firenzeRgb.jpg'
#I = im2double(rgb2gray(totensor(Image.open(requests.get(url, stream=True).raw))))

plt.figure(1)
plt.imshow(topilimg(I),cmap='gray')
plt.title('Original')
plt.show()
plt.figure(2)
imhist(I)
plt.show()

#%% ヒストグラム均等化前の加工
# (Process before histogram equalization)
# 
# $$v=T_1(x)= x^\gamma $$
# 
# $$y=T_2(v)= \frac{1}{2}(\mathrm{sign}(2v-1)|2v-1|^{10^{-\alpha}}+1)$$

# Definition of process
gamma = 1.0
alpha = 0.0
T1 = lambda x: torch.pow(x,gamma)
T2 = lambda v: 0.5*(torch.sign(2.0*v-1.0)*(torch.pow(torch.abs(2.0*v-1.0),10**(-alpha)))+1.0)
Tp = lambda x: T2(T1(x))
plt.figure(3)
plt.plot(locs,Tp(torch.tensor(locs)).detach().numpy())
plt.axis('square')
plt.grid()
plt.show()
# Preprocessing for histogram equalization
J = Tp(I)
plt.figure(4)
plt.imshow(topilimg(J),cmap='gray')
plt.title('Preprocessed image for histogram equalization')
plt.show()
plt.figure(5)
imhist(J)
plt.show()
#%% ヒストグラム均等化
# (Histgram equalization)

# Power law conversion with IMADUST function
K = totensor(ImageOps.equalize(topilimg(J)))
plt.figure(6)
plt.imshow(topilimg(K),cmap='gray')
plt.title('Result of histogram equalization')
plt.show()
plt.figure(7)
imhist(K)
plt.show()
#%% 
# © Copyright, Shogo MURAMATSU, All rights reserved.